### ISSUE TITLE
Support External Hazelcast instance (i.e. non-embedded)

### ISSUE THREAD

**Issue by sc-moonlight** [spring-projects/spring-session#339]
I could be missing something, but here's what I'm seeing.  I'm not sure if this can really be fixed, or if it just needs to be documented better that HazelcastHttpSessionConfiguration does not work for out of process Hazelcast server.

I enabled HazelcastHttpSessionConfiguration, and it works fine in a local profile that uses an embedded Hazelcast Instance.  However, when I switch to a profile that connects to a standalone Hazelcast instance, I get an exception thrown.

I'm setting up my hazelcast instance with this api:

```
    public static HazelcastInstance newHazelcastClient(ClientConfig config);
```

The first line in HazelcastHttpSessionConfiguration.configureSessionMap()  is 

```
		MapConfig sessionMapConfig = hazelcastInstance.getConfig().getMapConfig(sessionMapName);
```

then getConfig() throws this exception:

```
Caused by: java.lang.UnsupportedOperationException: Client cannot access cluster config!
          at com.hazelcast.client.impl.HazelcastClientInstanceImpl.getConfig(HazelcastClientInstanceImpl.java:278) ~[hazelcast-client-3.5.4.jar:3.5.4]
        at com.hazelcast.client.impl.HazelcastClientProxy.getConfig(HazelcastClientProxy.java:78) ~[hazelcast-client-3.5.4.jar:3.5.4]
        at org.springframework.session.hazelcast.config.annotation.web.http.HazelcastHttpSessionConfiguration.configureSessionMap(HazelcastHttpSessionConfiguration.java:98) ~[spring-session-1.1.0.M1.jar:na]
        at org.springframework.session.hazelcast.config.annotation.web.http.HazelcastHttpSessionConfiguration.sessionRepository(HazelcastHttpSessionConfiguration.java:70) ~[spring-session-1.1.0.M1.jar:na]
```

**Comment by rwinch** [comment]
@sc-moonlight Why are you needing to access the the `MapConfig`? Can you post your entire Spring Session Configuration?

cc @shakuzen @manderson23

**Comment by artembilan** [comment]
@rwinch , @sc-moonlight tells exactly about this code: https://github.com/spring-projects/spring-session/blob/master/spring-session/src/main/java/org/springframework/session/hazelcast/config/annotation/web/http/HazelcastHttpSessionConfiguration.java#L98.

**Comment by rwinch** [comment]
@artembilan Sorry my mistake! I see it now. This appears to be a bug.

**Comment by artembilan** [comment]
Np, @rwinch !

Looks like we don't have any choice for the fix unless, `try ... catch() {UnsupportedOperationException}` on that:

```
public Config getConfig() {
      throw new UnsupportedOperationException("Client cannot access cluster config!");
}
```
:smile:

**Comment by rwinch** [comment]
If either @manderson23 or @shakuzen have time to look into this, I'd appreciate it. Otherwise, I will do my best to get a fix in RC1.

**Comment by manderson23** [comment]
@rwinch @shakuzen

From a brief look at the docs perhaps a simple instanceof check for a client proxy would allow us to skip accessing the Map configuration? We could then document the assumption that the timeouts are configured correctly in both spring session and the hazelcast server for this case.

I can probably look at this within the week.

**Comment by shakuzen** [comment]
@rwinch @manderson23 Oops. I can take a look at this sometime this week as well and see what we can do about the client-server topology case with Hazelcast.

**Comment by rwinch** [comment]
@shakuzen @manderson23 Only looking at this very quickly so please feel free to tell me I'm wrong....

I wonder if it makes sense to remove the `maxInactiveIntervalInSeconds` from `@ EnableHazelcastHttpSession ` since this will not work for a remote client. Instead, users are responsible for setting the expiration when they create their `HazelcastInstance`.

If we do an instanceof check I'm afraid it will cause last minute problems for users who leverage Hazelcast locally and then go to production with a remote instance (and may not notice that `maxInactiveIntervalInSeconds` is ignored for the remote instance).

Thoughts?

**Comment by manderson23** [comment]
@rwinch @shakuzen 

Don't we still need ``maxInactiveIntervalInSeconds`` to set as the default on ``MapSessionRepository`` at https://github.com/spring-projects/spring-session/blob/master/spring-session/src/main/java/org/springframework/session/hazelcast/config/annotation/web/http/HazelcastHttpSessionConfiguration.java#L75?

**Comment by rwinch** [comment]
@manderson23 Hmm...that is a good point. This may take a bit more investigation to get an ideal outcome.

**Comment by manderson23** [comment]
@rwinch @shakuzun

Could we rename the parameter to refer to default and only set it on the repository? Then document that it should match hazelcast configuration for correct behaviour.

**Comment by rwinch** [comment]
@manderson23 Thanks for the feedback. We also need to consider that users may modify the expiration time using something like:

```
ExpiringSession s = ...;
s.setMaxInactiveIntervalInSeconds(10);
```

This means we would need to ensure the object expires. I wonder if we could solve the problem by doing something like this:

```java

	static class ExpiringSessionMap implements Map<String, ExpiringSession> {
		private IMap<String,ExpiringSession> delegate;
		
		ExpiringSessionMap(IMap<String,ExpiringSession> delegate) {
			this.delegate = delegate;
		}
		public ExpiringSession put(String key, ExpiringSession value) {
			if(value == null) {
				return delegate.put(key, value);
			}
			return delegate.put(key, value, value.getMaxInactiveIntervalInSeconds(), TimeUnit.SECONDS);
		}

		public int size() {
			return delegate.size();
		}

		public boolean isEmpty() {
			return delegate.isEmpty();
		}

		public boolean containsKey(Object key) {
			return delegate.containsKey(key);
		}

		public boolean containsValue(Object value) {
			return delegate.containsValue(value);
		}

		public ExpiringSession get(Object key) {
			return delegate.get(key);
		}

		public ExpiringSession remove(Object key) {
			return delegate.remove(key);
		}

		public void putAll(Map<? extends String, ? extends ExpiringSession> m) {
			delegate.putAll(m);
		}

		public void clear() {
			delegate.clear();
		}

		public Set<String> keySet() {
			return delegate.keySet();
		}

		public Collection<ExpiringSession> values() {
			return delegate.values();
		}

		public Set<java.util.Map.Entry<String, ExpiringSession>> entrySet() {
			return delegate.entrySet();
		}

		public boolean equals(Object o) {
			return delegate.equals(o);
		}

		public int hashCode() {
			return delegate.hashCode();
		}		
	}
```

Then we could use:

```java
MapSessionRepository sessionRepository = new MapSessionRepository(new ExpiringSessionMap(this.sessionsMap))
```

That would allow us to remove updating the configuration with the expiration time.
Thoughts?

PS: I haven't verified this works, but it seems like it might.

**Comment by manderson23** [comment]
@rwinch At a glance that looks reasonable.

**Comment by rwinch** [comment]
@manderson23 Thanks for the feedback! Is this something you or @shakuzen would be interested in sending a PR for?

**Comment by manderson23** [comment]
@rwinch if @shakuzen agrees this is the way to go I could probably integrate it within the week.

**Comment by rwinch** [comment]
@manderson23 Awesome! I'd love a to see a PR for this. That will let me focus on some other things. I'm hoping to get an 1.1.0.RC1 out the week of Feb 1 :)

**Comment by shakuzen** [comment]
@manderson23 @rwinch 
I am thinking aloud here to go over it all and make sure we have all our bases covered.

TL;DR Sounds good as long as we update the documentation accordingly and remove the current option to set idle timeout in your Hazelcast configuration, which I believe won't work (if it is longer than 1800 seconds, anyways).

**Proposed solution** (as I understand it):
- Stop setting the timeout in the Hazelcast MapConfig at all
- Introduce delegated Map implementation that sets the expiration time for each Hazelcast entry based on the ExpiringSession's expiration time

**Cases we tried to cover** and how they hold up with the proposed solution:
- default expiration - defaults to 1800, set for `MapSessionRepository`
- "" configure expiration in Hazelcast outside `HazelcastHttpSessionConfiguration` - need to add a null check before calling `sessionRepository.setDefaultMaxInactiveInterval(maxInactiveIntervalInSeconds)`
Even if the above is not set, `MapSession` has a default expiration time of 1800 - **this is problematic**
- custom overall expiration - will be set for the `MapSessionRepository` which overwrites the default for individual `MapSession` at creation
- custom individual expiration - set on the fly after Session creation `s.setMaxInactiveIntervalInSeconds(10)` - `ExpiringSessionMap` handles when persisting

Thus, everything looks good except that it seems we need to **remove the option** to set the expiration time in your Hazelcast configuration (by setting `maxInactiveIntervalInSeconds` to "" with `@EnableHazelcastHttpSession`). Accordingly, the documentation and JavaDocs need to be updated. Any configuration regarding TTI in one's Hazelcast MapConfig will be essentially ignored because we are setting individual entry TTI on put.

In some ways, I do like that no special Hazelcast map configuration is required (or allowed, for that matter) for Spring Session with this solution. In other ways, it seems less than ideal to be setting the TTI value for every entry when the vast majority are expected to use the default or globally configured value - though I think at least performance-wise it is a non-issue.

Anyways, let me know if I've missed something.

PS: I wish we had a Gitter (or other alternative) to chat about things like this rather than just comments.

**Comment by manderson23** [comment]
@rwinch @shakuzen 

Removing the option to support an empty string is OK with me since I'm configuring the repository directly from Spring XML anyway. (Something in the back of my mind makes me think I suggested this not that it matters anyway!).

So who is going to fix? If the target is a release within the week I should be able to do it. But would also be happy for @shakuzen to take it.

**Comment by shakuzen** [comment]
@manderson23 

I am fine reviewing or making the changes. I don't think I will have time today (Monday Japan time), but I probably could take a stab at making the changes tomorrow (Tuesday Japan time). If you have availability Monday or Tuesday and want to make the changes, just let me know. Otherwise I'll try to work on it Tuesday. We might need a day or two for review and additional changes, so if we want it done this week, Tuesday is probably the latest we want to start.

**Comment by rwinch** [comment]
Thanks for volunteering and the thorough inspection @shakuzen! Your proposed changes sound good to me. At your suggestion I also added Gitter for spring-session https://github.com/spring-projects/spring-session/commit/207e4e27b3dc7022a62979126c9a5a9a78b005dc

**Comment by manderson23** [comment]
@sc-moonlight @artembilan 

There is a fix for this pending merge in my branch at https://github.com/manderson23/spring-session/tree/gh-339
It hasn't yet been tested with an external Hazelcast instance. If either of you were able to build and test to see if it solves the issue we would be very grateful.

**Comment by artembilan** [comment]
??? Not sure what is the problem to start statically `Hazelcast.newHazelcastInstance()` (embedded in your terms?) and use the bean like:
```
@Bean
public HazelcastInstance hazelcastClient() {
       return HazelcastClient.newHazelcastClient();
}
```
When the first one starts an external Hazelcast node and the second one just connects to the existing Hazelcast cluster over `localhost` hook.

And we just covers the ` external Hazelcast instance` case in the one test!

Or am I missing anything?

**Comment by sc-moonlight** [comment]
Starting statically works.  It's when passing a ClientConfig , that it has a problem.  In my situation, I'm setting it up to talk to a HazelcastCluster that does not include localhost.

**Comment by artembilan** [comment]
>  It's when passing a ClientConfig , that it has a problem.

Sorry, I am really missing anything here...

What is the problem do you have there? Please, share.

According to the `ExpiringSessionMap` you have just only this code around TTL:
```
return delegate.put(key, value, value.getMaxInactiveIntervalInSeconds(), TimeUnit.SECONDS);
```
Which in case of `ClientMapProxy` impl is like:
```
protected V putInternal(long ttl, TimeUnit timeunit, Data keyData, Data valueData) {
        ClientMessage request = MapPutCodec.encodeRequest(name, keyData, valueData,
                ThreadUtil.getThreadId(), getTimeInMillis(ttl, timeunit));
        ClientMessage response = invoke(request, keyData);
        MapPutCodec.ResponseParameters resultParameters = MapPutCodec.decodeResponse(response);
        return toObject(resultParameters.response);
    }
```
So, we're fine with the new `HazelcastHttpSessionConfiguration` fix and don't have any access from the `HazelcastClient` to any critical config portions.

That's why I really don't see reason to worry about remote Hazelcast for our test-case, because the `HazelcastClient` impl is the same.

**Comment by manderson23** [comment]
@artembilan The test isn't meant to cover an external Hazelcast instance. Unit and integration tests currently only use embedded instances for simplicity.

We were just looking to see if the original reporter and those who commented wanted to test (or review) to ensure that it covers their cases. We're reasonably confident it does but both myself and @shakuzen don't use `HazelcastClient` and haven't had the change to test in that setup.

**Comment by artembilan** [comment]
OK. Thank you for explaining your position.

Well, why just don't add integration test with the `HazelcastClient` variant instead of `embeddedHazelcast`:

````java
@BeforeClass
public static void setup() {
      Hazelcast.newHazelcastInstance();
}

@AfterClass
public static void teardown() {
      Hazelcast.shutdownAll();
}
....
@Configuration
@EnableHazelcastHttpSession
static class HazelcastClientSessionConfig {

     @Bean
     public HazelcastInstance hazelcastClient() {
           return HazelcastClient.newHazelcastClient();
     }

}
````
?

**Comment by rwinch** [comment]
Thanks for the original report @sc-moonlight! Thanks for all the work on this @manderson23 @shakuzen! Also thanks for the suggestion @artembilan on the test! Finally, thanks to @vpavic for providing the integration test in 7c53558454a160e0443d73366ecc9e0d32211c3f!

This issue was really a team effort :)

This is now in master at f2443f5e21b64c193cd2c1e4c4d27d1f9f7b20ce
